### PR TITLE
AO Detail Map

### PR DESCRIPTION
**Pull request by Mistale**
Fixes [#4807](https://github.com/playcanvas/engine/issues/4807)
Uses the concepts introduced by [#1968](https://github.com/playcanvas/engine/pull/1968).

PR adds the following properties to `pc.StandardMaterial` with documentation and tests:

* `aoDetailMap` (texture)
* `aoDetailMapChannel` ("r", "g", "b", "a", default is "g")
* `aoDetailMapUv` (number)
* `aoDetailMapTiling` (vec2)
* `aoDetailMapOffset` (vec2)
* `aoDetailMode` (`pc.DETAILMODE_*`)

When combined with a baked `aoMap`, this PR allows for high- and low-frequency AO to be mixed without requiring excessively large baked textures. Typical use-cases may be a fence, a car grille or similar where one might want to use a baked AO texture for the entire object, but use a tiling baked AO texture for the intricate details.

### Example:
image (3)
> From left to right: baked AO, baked AO + detail Normal, baked AO + detail Normal + detail AO 

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by slimbuck** [comment]
Hi @Mistale,

Sorry for our slow response on your PR, it's awesome so we'll review/test/merge real soon.

Thanks!!

File: src/scene/materials/standard-material.js
Comment (mvaligursky) [inline src/scene/materials/standard-material.js:473]: This should mention the default value is 'g'

**Review by mvaligursky — APPROVED** [review]
I'm approving this, with a small JSDocs suggestion, but would like @GSterbrant to have a look as well.

My question is regarding the aoDetailMode parameter - did you find this useful when testing, versus just using one fixed way to blend those ao maps?

File: src/scene/materials/standard-material.js
Comment (Maksims) [reply src/scene/materials/standard-material.js:473]: Is it common to use default value 'g' instead or 'r' in other similar properties?

Comment (GSterbrant) [reply src/scene/materials/standard-material.js:473]: We don't seem to have that mentioned elsewhere in the docs, but we do set 'g' as the default channel in the material. It should be down to how textures are packed, surely, unless there is some bit depth favor for green channels (which would make sense perceptually) for certain file formats?

**Review by GSterbrant — APPROVED** [review]
Looks good! Just remember to update the chunk API version to 1.63 before merge. 

It does spawn a conversation I think we've had before on the team which is that perhaps we should add details maps as an intrinsic engine feature instead of having to add a bunch of extra parameters manually for each texture you'd want to detail map.

File: src/scene/shader-lib/chunks/standard/frag/ao.js
Comment (mvaligursky) [reply src/scene/shader-lib/chunks/standard/frag/ao.js:7]: yep, love the idea of automatic optional detail map for any texture. Please create an issue for this @GSterbrant

File: src/scene/materials/standard-material.js
Comment (Mistale) [reply src/scene/materials/standard-material.js]: I've added a clarification about the default value. I can't figure out why 'g' is the default for ao either, but I figured I'd just follow the same standard as  `aoMapChannel`. 
When packing ORM-textures for glTF, "r" is used for occlusion. But it may be related to inherent precision issues as @GSterbrant mentions.

**Comment by Mistale** [comment]
> I'm approving this, with a small JSDocs suggestion, but would like @GSterbrant to have a look as well.
> 
> My question is regarding the aoDetailMode parameter - did you find this useful when testing, versus just using one fixed way to blend those ao maps?

I think that using `DETAILMODE_MUL` as default covers a lot of normal use-cases, with one specific exception:
- When multiplying primary and secondary AO together using mul, the results will always vary (slightly) by the pattern of the secondary map, even in areas where the primary occlusion should be dark enough to completely remove details.
In this case a choice between `DETAILMODE_MIN` and `DETAILMODE_MAX` can make the results "flatter" while at the same time allowing the artist to introduce a bias towards the lighter or darker spectrum.

The other options are mainly there for artistic freedom.

File: src/scene/shader-lib/chunks/chunk-validation.js
Comment (Mistale) [reply src/scene/shader-lib/chunks/chunk-validation.js:8]: Just wanted to check if this was intended for me or internally (for the team)? It's my first PR, so I'm not quite sure about how you usually handle some of the details 😃

Comment (slimbuck) [reply src/scene/shader-lib/chunks/chunk-validation.js:8]: Also just to add this is a great first PE @Mistale - thanks so much!

**Comment by willeastcott** [comment]
This now has two approvals - any objections to merging @mvaligursky and @slimbuck? Last call for comments! ⌛

**Comment by mvaligursky** [comment]
Lets merge it in.

**Pull request merged**
